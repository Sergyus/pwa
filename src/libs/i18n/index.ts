import i18n from 'i18next';
import { initReactI18next } from 'react-i18next';
import detector from 'i18next-browser-languagedetector';
import XHR from 'i18next-xhr-backend';

// import en from '../../../public/locales/en/en.json';
// import ru from '../../../public/locales/ru/ru.json';
// import uk from '../../../public/locales/uk/uk.json';

export const LANGUAGES = ['en', 'ru', 'uk'];

export default i18n
  .use(detector)
  .use(XHR)
  .use(initReactI18next)
  .init({
    debug: false,
    resources: {
      en: {
        translation: {
          React: 'Welcome to React and react-i18next',
        },
        menu: {
          home: 'Home',
          blog: 'Blog',
          about: 'About',
        },
      },
      ru: {
        translation: {
          React: 'Добро пожаловать в React и react-i18next',
        },
        menu: {
          home: 'Главная',
          blog: 'Блог',
          about: 'О Нас',
        },
      },
      uk: {
        translation: {
          React: 'Ласкаво просимо до React і response-i18next',
        },
        menu: {
          home: 'Головна',
          blog: 'Блог',
          about: 'Про нас',
        },
      },
    },
    lng: 'en',
    fallbackLng: 'en',
    keySeparator: false,
    interpolation: {
      escapeValue: false, // react already safes from xss
    },
    backend: {
      loadPath: '/locales/{{lng}}/{{ns}}.json',
    },
  });
