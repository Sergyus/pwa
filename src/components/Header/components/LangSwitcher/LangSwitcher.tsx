import React, { useState } from 'react';
import Menu from '@material-ui/core/Menu';
import MenuItem from '@material-ui/core/MenuItem';
import IconButton from '@material-ui/core/IconButton';
import i18n from 'i18next';
import { LANGUAGES } from '@libs/i18n';

export default function LangSwitcher(): JSX.Element {
  const [anchorLangEl, setAnchorLangEl] = useState<null | HTMLElement>(null);
  const menuId = 'Lang-switcher';
  const isMenuOpen = Boolean(anchorLangEl);

  function handleMenuOpen(event: React.MouseEvent<HTMLElement>) {
    setAnchorLangEl(event.currentTarget);
  }

  function handleMenuClose() {
    setAnchorLangEl(null);
  }

  async function changeLanguage(event: React.MouseEvent<HTMLElement>) {
    const lang = (event.target as Element).getAttribute('value');
    if (lang) {
      await i18n.changeLanguage(lang);
    }
    handleMenuClose();
  }

  return (
    <div style={{ marginRight: '10px' }}>
      <IconButton
        edge="end"
        aria-label="account of current user"
        aria-controls={menuId}
        aria-haspopup="true"
        onClick={handleMenuOpen}
        color="inherit"
      >
        <div style={{ fontSize: '18px' }}>
          {i18n.language.toLocaleUpperCase()}
        </div>
      </IconButton>

      <Menu
        anchorEl={anchorLangEl}
        anchorOrigin={{ vertical: 'top', horizontal: 'left' }}
        id={menuId}
        keepMounted
        transformOrigin={{ vertical: 'top', horizontal: 'left' }}
        open={isMenuOpen}
        onClose={handleMenuClose}
      >
        {LANGUAGES.map((lang) => (
          <MenuItem value={lang} key={lang} onClick={changeLanguage}>
            {lang}
          </MenuItem>
        ))}
      </Menu>
    </div>
  );
}
