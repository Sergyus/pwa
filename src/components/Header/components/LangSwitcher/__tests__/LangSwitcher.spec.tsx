import React from 'react';
import { render, screen } from '@testing-library/react';
import LangSwitcher from '@components/Header/components/LangSwitcher';
import userEvent from '@testing-library/user-event';
import i18n from 'i18next';

describe('<LangSwitcher />', () => {
  it('renders <LangSwitcher />', () => {
    const { asFragment } = render(<LangSwitcher />);
    const btn = screen.getByTestId('langBtn');
    expect(btn).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('should have correct lang in button', () => {
    render(<LangSwitcher />);
    const lang = i18n.language.toLocaleUpperCase();
    const btn = screen.getByTestId('langBtn');
    expect(btn).toHaveTextContent(lang);
  });

  it('should render menu on click', () => {
    render(<LangSwitcher />);
    const btn = screen.getByTestId('langBtn');
    const menu = screen.getByTestId('langMenu');
    userEvent.click(btn);
    expect(menu).toBeVisible();
  });

  // it('should have all select items', () => {
  //   render(<LangSwitcher />);
  //   const menu = screen.getByTestId('langMenu');
  // });

  // it('should be change lang on select', () => {
  //   render(<LangSwitcher />);
  //   const lang = i18n.language.toLocaleUpperCase();
  //   const btn = screen.getByTestId('langBtn');
  //   const langItems = screen.getAllByTestId('langItem');
  //   // const changeLanguage = jest.fn();
  //
  //   userEvent.click(langItems);
  //
  //   expect(btn).toHaveTextContent(lang);
  //   // expect(changeLanguage).toHaveBeenCalledTimes(0);
  // });
});
