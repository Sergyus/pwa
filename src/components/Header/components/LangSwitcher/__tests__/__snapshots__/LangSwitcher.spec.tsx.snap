// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LangSwitcher /> renders <LangSwitcher /> 1`] = `
<DocumentFragment>
  <div
    style="margin-right: 10px;"
  >
    <button
      aria-controls="Lang-switcher"
      aria-haspopup="true"
      aria-label="account of current user"
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeEnd"
      data-testid="langBtn"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiIconButton-label"
      >
        <div
          style="font-size: 18px;"
        >
          EN
        </div>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</DocumentFragment>
`;
